Resolve alternative names

keszybz opened the pull request:
Minor follow-up for #14365.

poettering commented on src/network/networkctl.c:
```diff
@@ -60,6 +60,21 @@ static bool arg_legend = true;
+
+        r = rtnl_resolve_link_alternative_name(rtnl, name);
+        if (r > 0)
+                return r;
```
hmm, shouldn't we check this already in parse_ifindex_or_ifname() itself?

i.e. introduce a if_nametoindex_harder() or so that is like if_nametoindex() but also checks alternative names? I mean, it's a pity the kernel doesn't support this natively in SIOCGIFINDEX but I think we should just add our own helper tha tdoes this anyway?

i.e. I am pretty sure we should just support alternative names everywhere, and not just in networkctl?

poettering commented:
lgtm, but I think this isn't going far enough, see above. We should support the alternative names everywhere, and that means all direct and indirect uses of if_nametoindex() should be updated to support alternative names, via some wrapper if_nametoindex_harder()

keszybz commented:
Damn, this is harder than it seems. Resolution of alternate names requires sd-netlink, so it shouldn't be used in src/basic/. And there are uses of if_nametoindex() that should be "upgraded" to support alternate names in src/basic/.

keszybz commented:
Updated to "tree-wide". PTAL.

yuwata commented on src/nspawn/nspawn-network.c:
```diff
@@ -492,7 +484,7 @@ int test_network_interface_initialized(const char *name) {
         /* udev should be around. */
 
-        ifi = parse_interface(name);
+        ifi = resolve_ifname(NULL, name);
```
`resolve_ifname_or_warn()`

keszybz replied:
Good catch, but I think the old code was doing double logging. `join_bridge` is only called from `setup_bridge`, and that prints the error. We'll get something like ""Failed to add interface … to bridge …: no such device", which I think is appropriate.

yuwata commented on src/nspawn/nspawn-network.c:
```diff
@@ -532,7 +524,7 @@ int move_network_interfaces(int netns_fd, char **ifaces) {
                 int ifi;
 
-                ifi = parse_interface(*i);
+                ifi = resolve_ifname(&rtnl, *i);
```
ditto.

yuwata commented on src/nspawn/nspawn-network.c:
```diff
@@ -571,7 +563,7 @@ int setup_macvlan(const char *machine_name, pid_t pid, char **ifaces) {
                 int ifi;
 
-                ifi = parse_interface(*i);
+                ifi = resolve_ifname(&rtnl, *i);
```
ditto.

yuwata commented on src/nspawn/nspawn-network.c:
```diff
@@ -657,7 +649,7 @@ int setup_ipvlan(const char *machine_name, pid_t pid, char **ifaces) {
                 int ifi;
 
-                ifi = parse_interface(*i);
+                ifi = resolve_ifname(&rtnl, *i);
```
ditto.

keszybz commented:
I reworked the patchset quite a bit. In the previous version, there was just a single parsing function, that would always resolve either an ifindex or the ifname or the alternative name. This made things simpler, but it would resolve an ifindex in some places where we probably shouldn't. I reworked the patch to not do this, so there's a bit more code, but less change in behaviour. Now the change is only that resolution of alt ifnames is added whereever ifnames were parsed.
